stablehlo/reference/Interpreter.cpp

```diff
@@ -35,142 +36,174 @@ llvm::Expected<SmallVector<Tensor>> eval(func::FuncOp func,
     return invalidArgument("Expected one region in func %s",
                            func.getName().str().c_str());
 
-  if (!func.getBody().hasOneBlock())
-    return invalidArgument("Expected one block in func %s",
-                           func.getName().str().c_str());
-
   Block &block = func.front();
   if (block.getNumArguments() != args.size())
     return invalidArgument(
         "Expected same amount of func arguments in %s "
         "and runtime arguments (%d)",
         func.getName().str().c_str(), args.size());
 
-  llvm::DenseMap<Value, Tensor> stackFrame;
-  for (auto [ssaArg, runtimeArg] : llvm::zip(block.getArguments(), args))
-    stackFrame[ssaArg] = runtimeArg;
+  return eval(func.getBody(), args, nullptr);
+}
+
+llvm::Expected<SmallVector<Tensor>> eval(
+    Region &region, ArrayRef<Tensor> runtimeArgs,
+    const InterpreterScope *const parentScope) {
+  if (!region.hasOneBlock())
+    return invalidArgument("Expected single block region");
+
+  InterpreterScope scope(parentScope);
+
+  Block &block = region.front();
+  if (block.getArguments().size() != runtimeArgs.size())
+    report_fatal_error(invalidArgument(
+        "Expected same amount of block arguments and runtime arguments (%d)",
+        runtimeArgs.size()));
+
+  for (auto [ssaArg, runtimeArg] :
+       llvm::zip(block.getArguments(), runtimeArgs)) {
+    assert(ssaArg.getType() == runtimeArg.getType());
+    scope.add(ssaArg, runtimeArg);
+  }
+
+  auto fetchOperand = [&](Value value) -> Tensor { return scope.find(value); };
+
+  auto fetchVariadicOperands = [&](OperandRange values) {
+    return llvm::to_vector(llvm::map_range(
+        values, [&](Value value) { return fetchOperand(value); }));
+  };
 
   for (Operation &op : block) {
-    auto fetchOperand = [&](Value value) -> Tensor {
-      auto it = stackFrame.find(value);
-      if (it != stackFrame.end()) return it->second;
-      llvm::report_fatal_error("Expected a terminator when evaluating func");
-    };
-    auto populateResults = [&](ArrayRef<Tensor> runtimeValues) {
-      assert(op.getNumResults() == runtimeValues.size());
+    auto addOpResultsToScope = [&](ArrayRef<Tensor> runtimeResults) {
+      assert(op.getNumResults() == runtimeResults.size());
       for (auto [ssaResult, runtimeResult] :
-           llvm::zip(op.getResults(), runtimeValues))
-        stackFrame[ssaResult] = runtimeResult;
+           llvm::zip(op.getResults(), runtimeResults)) {
+        if (ssaResult.getType() != runtimeResult.getType()) {
+          llvm::report_fatal_error(
+              "Expected same value for an SSA register and its evalted value");
+        }
+        scope.add(ssaResult, runtimeResult);
+      }
     };
```

Comment:
I think we should delete all the lambdas above and then introduce and use `InterpreterScope` methods instead (see comments to `InterpreterScope.h`).